In [1]:
import pandas as pd
import seaborn as sns
from matplotlib import pyplot as plt

In [2]:
all_cards = pd.read_csv('cleaned_datasets/allsets_cleaned.csv', index_col=0)

In [3]:
all_cards.head()

name colors  manaValue   manaCost  power  toughness  \
0  Aerial Modification      W        5.0     {4}{W}    NaN        NaN   
1     Aeronaut Admiral      W        4.0     {3}{W}    3.0        1.0   
2     Aether Inspector      W        4.0     {3}{W}    2.0        3.0   
3    Aethergeode Miner      W        2.0     {1}{W}    3.0        1.0   
4    Airdrop Aeronauts      W        5.0  {3}{W}{W}    4.0        3.0   

   loyalty                                               text    rarity  \
0      NaN  Enchant creature or Vehicle\nAs long as enchan...  uncommon   
1      NaN          Flying\nVehicles you control have flying.  uncommon   
2      NaN  Vigilance\nWhen Aether Inspector enters the ba...    common   
3      NaN  Whenever Aethergeode Miner attacks, you get {E...      rare   
4      NaN  Flying\nRevolt — When Airdrop Aeronauts enters...  uncommon   

         keywords supertypes        types          subtypes  
0         Enchant        NaN  Enchantment              Aura  
1          Flying        NaN     Creature      Human, Pilot  
2       Vigilance        NaN     Creature  Dwarf, Artificer  
3             NaN        NaN     Creature      Dwarf, Scout  
4  Flying, Revolt        NaN     Creature      Dwarf, Scout

In [4]:
all_cards.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7125 entries, 0 to 7124
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   name        7125 non-null   object 
 1   colors      7125 non-null   object 
 2   manaValue   7125 non-null   float64
 3   manaCost    6617 non-null   object 
 4   power       3867 non-null   float64
 5   toughness   3877 non-null   float64
 6   loyalty     168 non-null    float64
 7   text        7025 non-null   object 
 8   rarity      7125 non-null   object 
 9   keywords    3399 non-null   object 
 10  supertypes  763 non-null    object 
 11  types       7125 non-null   object 
 12  subtypes    4636 non-null   object 
dtypes: float64(4), object(9)
memory usage: 779.3+ KB


NaN values in the `keywords` column are to be expected, as many cards do not have keywords.  We will replace these with empty strings.

In [5]:
all_cards['keywords'].fillna('', inplace=True)

I want to get a dataframe that includes only creatures, but first we have to deal with one edge case:  non-creature cards that contain the "Transform" keyword.  I do this by writing out a function that can extract the mana value for cards that are formatted in the following way: "{B}{B}{2}: Transform".  We find the Transform keyword in the text of the card, then grab the string that immediately precedes it, calculate the additional cost to transform the card and add it to the correct cell (which will, in all cases, be the row immediately following the card with the Transform keyword.  This also fixes the mana value for the back sides of modal cards with the Transform keyword that are both creatures.

In [6]:
all_cards['text'].fillna('', inplace=True)

In [7]:
all_cards.loc[all_cards.keywords.str.contains('Transform', case=False)]

name colors  manaValue   manaCost  power  toughness  \
1546     Extricator of Sin      W        3.0     {2}{W}    0.0        3.0   
1557            Lone Rider      W        2.0     {1}{W}    1.0        1.0   
1579    Curious Homunculus      U        2.0     {1}{U}    1.0        1.0   
1582  Docent of Perfection      U        5.0  {3}{U}{U}    5.0        4.0   
1590       Grizzled Angler      U        3.0     {2}{U}    2.0        3.0   
...                    ...    ...        ...        ...    ...        ...   
6809     Primal Wellspring      L        4.0        NaN    NaN        NaN   
6814     Thaumatic Compass      C        2.0        {2}    NaN        NaN   
6815      Spires of Orazca      L        2.0        NaN    NaN        NaN   
6816          Treasure Map      C        2.0        {2}    NaN        NaN   
6817         Treasure Cove      L        2.0        NaN    NaN        NaN   

      loyalty                                               text    rarity  \
1546      NaN  When Extricator of Sin enters the battlefield,...  uncommon   
1557      NaN  First strike, lifelink\nAt the beginning of th...  uncommon   
1579      NaN  {T}: Add {C}. Spend this mana only to cast an ...  uncommon   
1582      NaN  Flying\nWhenever you cast an instant or sorcer...      rare   
1590      NaN  {T}: Mill two cards. Then if there is a colorl...  uncommon   
...       ...                                                ...       ...   
6809      NaN  (Transforms from Primal Amulet.)\n{T}: Add one...      rare   
6814      NaN  {3}, {T}: Search your library for a basic land...      rare   
6815      NaN  (Transforms from Thaumatic Compass.)\n{T}: Add...      rare   
6816      NaN  {1}, {T}: Scry 1. Put a landmark counter on Tr...      rare   
6817      NaN  (Transforms from Treasure Map.)\n{T}: Add {C}....      rare   

                               keywords supertypes     types        subtypes  
1546                Delirium, Transform        NaN  Creature   Human, Cleric  
1557  First strike, Lifelink, Transform        NaN  Creature   Human, Knight  
1579                          Transform        NaN  Creature      Homunculus  
1582                  Flying, Transform        NaN  Creature  Insect, Horror  
1590                    Mill, Transform        NaN  Creature           Human  
...                                 ...        ...       ...             ...  
6809                          Transform        NaN      Land             NaN  
6814                          Transform        NaN  Artifact             NaN  
6815                          Transform        NaN      Land             NaN  
6816                    Scry, Transform        NaN  Artifact             NaN  
6817                          Transform        NaN      Land             NaN  

[197 rows x 13 columns]

In [8]:
def backwards_extract (df, keyword):
    import re
    
    # store the relevant index numbers from the dataframe
    indexes = df.loc[df.text.str.contains(keyword)].index

    # loop over each element to be changed
    for count, text in enumerate(df.loc[df.text.str.contains(keyword)].text):

        # find occurence of keyword
        c = text.find(keyword)

        # capture the text before the keyword
        if c <= 12:
            text_value = text[:c]
        else:
            text_value = text[c-12:c-2]

        # check to see that a curly brace exists before the keyword, confirming
        # that the text directly before it indicates an additional mana cost
        if '}' in text_value:
            text_value = text_value[text_value.find('{'):]

            # remove curly braces, Tap symbols, whitespace, colons
            text_value = re.sub('[,T:\s}{]', '', text_value)            

            # calculate value to use in dataframe 
            new_value = len(text_value) 
            if text_value[0].isdigit():
                new_value += int(text_value[0]) - 1

            # update the value of the NEXT card in the dataframe, the one that
            # our identified cards transform into
            df.iat[indexes[count] + 1, 2] += new_value
            
            # print a list of changes made so we can confirm we got it right
            print(df.iloc[indexes[count] +1, :4])

In [9]:
backwards_extract(all_cards, 'Transform')

name         Conduit of Emrakul
colors                        C
manaValue                   8.0
manaCost                    NaN
Name: 1655, dtype: object
name         Erupting Dreadwolf
colors                        C
manaValue                  10.0
manaCost                    NaN
Name: 1675, dtype: object
name         Dronepack Kindred
colors                       C
manaValue                 12.0
manaCost                   NaN
Name: 1682, dtype: object
name         Sinuous Predator
colors                      C
manaValue                 6.0
manaCost                  NaN
Name: 1698, dtype: object
name         Howling Chorus
colors                    C
manaValue               9.0
manaCost                NaN
Name: 1703, dtype: object
name         Fibrous Entangler
colors                       C
manaValue                 11.0
manaCost                   NaN
Name: 1710, dtype: object
name         Ulvenwald Abomination
colors                           C
manaValue                      9.0
man

In [10]:
creatures = all_cards.loc[all_cards.types.str.contains('Creature')].copy()
creatures.reset_index(drop=True, inplace=True)

In [11]:
creatures.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3845 entries, 0 to 3844
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   name        3845 non-null   object 
 1   colors      3845 non-null   object 
 2   manaValue   3845 non-null   float64
 3   manaCost    3698 non-null   object 
 4   power       3814 non-null   float64
 5   toughness   3824 non-null   float64
 6   loyalty     0 non-null      float64
 7   text        3845 non-null   object 
 8   rarity      3845 non-null   object 
 9   keywords    3845 non-null   object 
 10  supertypes  457 non-null    object 
 11  types       3845 non-null   object 
 12  subtypes    3845 non-null   object 
dtypes: float64(4), object(9)
memory usage: 390.6+ KB


In [12]:
creatures.head()

name colors  manaValue   manaCost  power  toughness  \
0       Aeronaut Admiral      W        4.0     {3}{W}    3.0        1.0   
1       Aether Inspector      W        4.0     {3}{W}    2.0        3.0   
2      Aethergeode Miner      W        2.0     {1}{W}    3.0        1.0   
3      Airdrop Aeronauts      W        5.0  {3}{W}{W}    4.0        3.0   
4  Audacious Infiltrator      W        2.0     {1}{W}    3.0        1.0   

   loyalty                                               text    rarity  \
0      NaN          Flying\nVehicles you control have flying.  uncommon   
1      NaN  Vigilance\nWhen Aether Inspector enters the ba...    common   
2      NaN  Whenever Aethergeode Miner attacks, you get {E...      rare   
3      NaN  Flying\nRevolt — When Airdrop Aeronauts enters...  uncommon   
4      NaN  Audacious Infiltrator can't be blocked by arti...    common   

         keywords supertypes     types          subtypes  
0          Flying        NaN  Creature      Human, Pilot  
1       Vigilance        NaN  Creature  Dwarf, Artificer  
2                        NaN  Creature      Dwarf, Scout  
3  Flying, Revolt        NaN  Creature      Dwarf, Scout  
4                        NaN  Creature      Dwarf, Rogue

The column "Loyalty" is a value that is unique to planeswalkers, so we're going to drop that right away.

In [13]:
creatures.drop(['loyalty'], axis=1, inplace=True)

In [14]:
creatures.head()

name colors  manaValue   manaCost  power  toughness  \
0       Aeronaut Admiral      W        4.0     {3}{W}    3.0        1.0   
1       Aether Inspector      W        4.0     {3}{W}    2.0        3.0   
2      Aethergeode Miner      W        2.0     {1}{W}    3.0        1.0   
3      Airdrop Aeronauts      W        5.0  {3}{W}{W}    4.0        3.0   
4  Audacious Infiltrator      W        2.0     {1}{W}    3.0        1.0   

                                                text    rarity  \
0          Flying\nVehicles you control have flying.  uncommon   
1  Vigilance\nWhen Aether Inspector enters the ba...    common   
2  Whenever Aethergeode Miner attacks, you get {E...      rare   
3  Flying\nRevolt — When Airdrop Aeronauts enters...  uncommon   
4  Audacious Infiltrator can't be blocked by arti...    common   

         keywords supertypes     types          subtypes  
0          Flying        NaN  Creature      Human, Pilot  
1       Vigilance        NaN  Creature  Dwarf, Artificer  
2                        NaN  Creature      Dwarf, Scout  
3  Flying, Revolt        NaN  Creature      Dwarf, Scout  
4                        NaN  Creature      Dwarf, Rogue

In [15]:
len(creatures[creatures.manaCost.isna()])

147

There are 147 rows with NaN values for manaCost.  We have zero rows that have NaN values in manaValue, though, so these are almost definitely the backs of modal cards that transform through one game mechanic or another.  The amount of resources spent to cast the card is accurately reflected in the manaValue column, though the conditions for flipping the card are not going to be considered at this stage of our investigation.  Still, we'll do the same as we did above and check that this is true of about 10% of the data before dropping the column to make sure we didn't inadvertently introduce any errors.

In [16]:
creatures[creatures.manaCost.isna()].sample(15)

name colors  manaValue manaCost  power  toughness  \
2848        Wayward Disciple      B        3.0      NaN    2.0        4.0   
2937            Skin Shedder      R        1.0      NaN    3.0        4.0   
1563         Seasoned Cathar      W        5.0      NaN    3.0        3.0   
1639            Awoken Demon      B        1.0      NaN    4.0        4.0   
2858          Perfected Form      U        4.0      NaN    5.0        4.0   
906        Fibrous Entangler      C       11.0      NaN    4.0        6.0   
1736  Olag, Ludevic's Hubris   B, U        2.0      NaN    4.0        4.0   
2200   Hand of Enlightenment      W        2.0      NaN    2.0        2.0   
3383        Lambholt Ravager      R        4.0      NaN    4.0        4.0   
1727     Departed Soulkeeper   U, W        2.0      NaN    3.0        1.0   
1596      Hook-Haunt Drifter      U        2.0      NaN    1.0        2.0   
828         Voracious Reader      C        2.0      NaN    3.0        4.0   
3284        Cackling Culprit      B        2.0      NaN    3.0        5.0   
3457       Forsaken Thresher      C        3.0      NaN    5.0        5.0   
3400  Hollowhenge Huntmaster      G        6.0      NaN    6.0        6.0   

                                                   text    rarity  \
2848  Whenever Wayward Disciple or another creature ...  uncommon   
2937                                                     uncommon   
1563                                           Lifelink  uncommon   
1639                                                       common   
2858                                             Flying  uncommon   
906   Vigilance\nFibrous Entangler must be blocked i...  uncommon   
1736  As this creature transforms into Olag, Ludevic...      rare   
2200                                       First strike    common   
3383  Whenever you cast a noncreature spell, Lambhol...  uncommon   
1727  Flying\nDeparted Soulkeeper can block only cre...  uncommon   
1596  Flying\nIf Hook-Haunt Drifter would be put int...    common   
828   Prowess (Whenever you cast a noncreature spell...  uncommon   
3284  Whenever Cackling Culprit or another creature ...  uncommon   
3457  At the beginning of your precombat main phase,...  uncommon   
3400  Hexproof\nOther permanents you control have he...    mythic   

                  keywords supertypes                  types  \
2848                              NaN               Creature   
2937                              NaN               Creature   
1563              Lifelink        NaN               Creature   
1639                              NaN               Creature   
2858                Flying        NaN               Creature   
906              Vigilance        NaN               Creature   
1736             Transform  Legendary               Creature   
2200          First strike        NaN  Enchantment, Creature   
3383            Nightbound        NaN               Creature   
1727                Flying        NaN               Creature   
1596                Flying        NaN               Creature   
828                Prowess        NaN               Creature   
3284                              NaN               Creature   
3457                              NaN     Artifact, Creature   
3400  Hexproof, Nightbound        NaN               Creature   

                 subtypes  
2848        Human, Cleric  
2937       Insect, Horror  
1563        Human, Knight  
1639                Demon  
2858       Insect, Horror  
906     Eldrazi, Werewolf  
1736               Zombie  
2200          Human, Monk  
3383             Werewolf  
1727               Spirit  
1596               Spirit  
828   Eldrazi, Homunculus  
3284         Human, Rogue  
3457            Construct  
3400             Werewolf

In checking to make sure I didn't introduce any errors, I realized that the listed Mana Values for cards with the "Disturb" keyword were not reflective of the cost to cast their disturb mode, so I went ahead and built out a function that will let us extract the Disturb cost from the card text and replace that in the mana value field.

In [17]:
def extract_mana_value (df, keyword):
    import re
    
    # store the relevant index numbers from the dataframe
    indexes = df.loc[df.text.str.contains(keyword)].index

    # loop over each element to be changed
    for count, text in enumerate(df.loc[df.text.str.contains(keyword)].text):

        # find occurence of keyword
        c = text.find(keyword)

        # capture the text after the keyword that lists the mana cost of casting 
        # the card, leaving out leading and trailing curly braces.
        text_value = text[c+9:text.rfind('}', c+8, c+20)]

        # remove the rest of the curly braces
        text_value = re.sub('[,T:\s}{]', '', text_value)

        # calculate value to use in dataframe 
        new_value = len(text_value) 
        if text_value[0].isdigit():
            new_value += int(text_value[0]) - 1

        # update the value in the dataframe
        df.iat[indexes[count], 2] = new_value

In [18]:
extract_mana_value(creatures, 'Disturb')

In [19]:
creatures.loc[creatures.text.str.contains('Disturb')].head()

name colors  manaValue manaCost  power  toughness  \
1564    Beloved Beggar      W        6.0   {1}{W}    0.0        4.0   
1573  Chaplain of Alms      W        4.0      {W}    1.0        1.0   
1584   Lunarch Veteran      W        2.0      {W}    1.0        1.0   
1586   Mourning Patrol      W        4.0   {2}{W}    2.0        3.0   
1595   Baithook Angler      U        2.0   {1}{U}    2.0        1.0   

                                                   text    rarity  \
1564  Disturb {4}{W}{W} (You may cast this card from...  uncommon   
1573  First strike\nWard {1} (Whenever this creature...  uncommon   
1584  Whenever another creature enters the battlefie...    common   
1586  Vigilance\nDisturb {3}{W} (You may cast this c...    common   
1595  Disturb {1}{U} (You may cast this card from yo...    common   

                                    keywords supertypes     types  \
1564                      Disturb, Transform        NaN  Creature   
1573  Disturb, First strike, Transform, Ward        NaN  Creature   
1584                      Disturb, Transform        NaN  Creature   
1586           Disturb, Transform, Vigilance        NaN  Creature   
1595                      Disturb, Transform        NaN  Creature   

            subtypes  
1564  Human, Peasant  
1573   Human, Cleric  
1584   Human, Cleric  
1586  Human, Soldier  
1595  Human, Peasant

We have two exceptions with cards that come into play as lands and then can later be flipped into creatures.  I'll go ahead and manually set reasonable manaValues for those two, since it's such a small number.

In [20]:
creatures[creatures.manaCost.isna()].query('manaValue == 0')

name colors  manaValue manaCost  power  toughness  \
1750               Creeping Inn      B        0.0      NaN    3.0        7.0   
2989  Ormendahl, Profane Prince      B        0.0      NaN    9.0        7.0   

                                                   text  rarity  \
1750  Whenever Creeping Inn attacks, you may exile a...  mythic   
2989            Flying, lifelink, indestructible, haste    rare   

                                     keywords supertypes               types  \
1750                                                 NaN  Artifact, Creature   
2989  Flying, Haste, Indestructible, Lifelink  Legendary            Creature   

               subtypes  
1750  Horror, Construct  
2989              Demon

In [21]:
creatures.iat[2989, 2] = 5
creatures.iat[1750, 2] = 3

In [22]:
creatures.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3845 entries, 0 to 3844
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   name        3845 non-null   object 
 1   colors      3845 non-null   object 
 2   manaValue   3845 non-null   float64
 3   manaCost    3698 non-null   object 
 4   power       3814 non-null   float64
 5   toughness   3824 non-null   float64
 6   text        3845 non-null   object 
 7   rarity      3845 non-null   object 
 8   keywords    3845 non-null   object 
 9   supertypes  457 non-null    object 
 10  types       3845 non-null   object 
 11  subtypes    3845 non-null   object 
dtypes: float64(3), object(9)
memory usage: 360.6+ KB


In [23]:
creatures[creatures.manaCost.isna()].sample(15)

name colors  manaValue manaCost  power  toughness  \
1635          Leeching Lurker      B        3.0      NaN    4.0        4.0   
2830     Avacyn, the Purifier      R        5.0      NaN    6.0        5.0   
2832      Lunarch Inquisitors      W        4.0      NaN    4.0        4.0   
1672    Harvesttide Assailant      R        3.0      NaN    4.0        4.0   
2922        Vildin-Pack Alpha      R        3.0      NaN    4.0        3.0   
1733    Lord of the Ulvenwald   G, R        2.0      NaN    3.0        3.0   
3284         Cackling Culprit      B        2.0      NaN    3.0        5.0   
3317       Cipherbound Spirit      U        2.0      NaN    3.0        2.0   
2913             Neck Breaker      R        3.0      NaN    4.0        3.0   
2257  Animus of Night's Reach      B        4.0      NaN    0.0        4.0   
2876     Persistent Nightmare      U        4.0      NaN    1.0        1.0   
3394   Volt-Charged Berserker      R        2.0      NaN    4.0        3.0   
836         Grisly Anglerfish      C        3.0      NaN    4.0        5.0   
2840     Westvale Cult Leader      W        2.0      NaN    NaN        NaN   
3336       Depraved Harvester      B        3.0      NaN    4.0        3.0   

                                                   text    rarity  \
1635  Lifelink\nNightbound (If a player casts at lea...      rare   
2830  Flying\nWhen this creature transforms into Ava...    mythic   
2832  When this creature transforms into Lunarch Inq...  uncommon   
1672  Trample\nNightbound (If a player casts at leas...    common   
2922  Whenever a Werewolf enters the battlefield und...      rare   
1733  Other Wolves and Werewolves you control get +1...  uncommon   
3284  Whenever Cackling Culprit or another creature ...  uncommon   
3317  Flying\nCipherbound Spirit can block only crea...  uncommon   
2913  Attacking creatures you control get +1/+0 and ...  uncommon   
2257  Menace (This creature can't be blocked except ...  uncommon   
2876  Skulk (This creature can't be blocked by creat...    mythic   
3394                Volt-Charged Berserker can't block.  uncommon   
836   {6}: Creatures your opponents control attack t...  uncommon   
2840  Westvale Cult Leader's power and toughness are...      rare   
3336                                           Lifelink    common   

                  keywords supertypes                  types          subtypes  
1635  Lifelink, Nightbound        NaN               Creature     Leech, Horror  
2830     Flying, Transform  Legendary               Creature             Angel  
2832             Transform        NaN               Creature     Human, Cleric  
1672   Nightbound, Trample        NaN               Creature          Werewolf  
2922             Transform        NaN               Creature          Werewolf  
1733            Nightbound        NaN               Creature          Werewolf  
3284                              NaN               Creature      Human, Rogue  
3317                Flying        NaN               Creature            Spirit  
2913             Transform        NaN               Creature          Werewolf  
2257                Menace        NaN  Enchantment, Creature            Spirit  
2876                 Skulk        NaN               Creature         Nightmare  
3394                              NaN               Creature  Human, Berserker  
836                               NaN               Creature     Eldrazi, Fish  
2840                              NaN               Creature     Human, Cleric  
3336              Lifelink        NaN               Creature     Human, Knight

At this point, we should have dealt with most or all of the incorrectly assigned manaValue creatures, though we are not accounting for non-mana costs.  Doing so would be significantly more complicated and goes beyond the scope of the current investigation, so I'm happy to move on to other things that need to be corrected.  At this point, we should not need the manaCost column any longer, so I'll drop it.

In [24]:
creatures.drop(columns='manaCost', inplace=True)

In [25]:
creatures.head()

name colors  manaValue  power  toughness  \
0       Aeronaut Admiral      W        4.0    3.0        1.0   
1       Aether Inspector      W        4.0    2.0        3.0   
2      Aethergeode Miner      W        2.0    3.0        1.0   
3      Airdrop Aeronauts      W        5.0    4.0        3.0   
4  Audacious Infiltrator      W        2.0    3.0        1.0   

                                                text    rarity  \
0          Flying\nVehicles you control have flying.  uncommon   
1  Vigilance\nWhen Aether Inspector enters the ba...    common   
2  Whenever Aethergeode Miner attacks, you get {E...      rare   
3  Flying\nRevolt — When Airdrop Aeronauts enters...  uncommon   
4  Audacious Infiltrator can't be blocked by arti...    common   

         keywords supertypes     types          subtypes  
0          Flying        NaN  Creature      Human, Pilot  
1       Vigilance        NaN  Creature  Dwarf, Artificer  
2                        NaN  Creature      Dwarf, Scout  
3  Flying, Revolt        NaN  Creature      Dwarf, Scout  
4                        NaN  Creature      Dwarf, Rogue

There is a small subset of creatures whose power and/or toughness is listed as '*'.  This means that those values are variable and dependent on game-state conditions.  We've only got 34 of these creatures in our dataset of almost 4,000, comprising less than 1% of the creatures.  Rather than spend an inordinate amount of time accounting for these edge cases, I choose to drop them from the dataframe.

In [26]:
len(creatures[creatures.power.isnull() | creatures.toughness.isnull()])

34

In [27]:
creatures.drop(creatures[creatures.power.isnull() | creatures.toughness.isnull()].index, inplace=True)
creatures.reset_index(drop=True, inplace=True)

In [28]:
creatures.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3811 entries, 0 to 3810
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   name        3811 non-null   object 
 1   colors      3811 non-null   object 
 2   manaValue   3811 non-null   float64
 3   power       3811 non-null   float64
 4   toughness   3811 non-null   float64
 5   text        3811 non-null   object 
 6   rarity      3811 non-null   object 
 7   keywords    3811 non-null   object 
 8   supertypes  446 non-null    object 
 9   types       3811 non-null   object 
 10  subtypes    3811 non-null   object 
dtypes: float64(3), object(8)
memory usage: 327.6+ KB


## To do: Move this Venture change to initial cleaning on combined_df

In [29]:
creatures[creatures.keywords.str.contains('Venture', case=False)].head()

name colors  manaValue  power  toughness  \
104         Cloister Gargoyle      W        3.0    0.0        4.0   
115         Keen-Eared Sentry      W        2.0    2.0        1.0   
119  Nadaar, Selfless Paladin      W        3.0    3.0        3.0   
121               Planar Ally      W        5.0    3.0        3.0   
123             Ranger's Hawk      W        1.0    1.0        1.0   

                                                  text    rarity  \
104  When Cloister Gargoyle enters the battlefield,...  uncommon   
115  You have hexproof. (You can't be the target of...  uncommon   
119  Vigilance\nWhenever Nadaar, Selfless Paladin e...      rare   
121  Flying\nWhenever Planar Ally attacks, venture ...    common   
123  Flying\n{3}, {T}, Tap another untapped creatur...    common   

                                keywords supertypes               types  \
104             Venture into the dungeon        NaN  Artifact, Creature   
115             Venture into the dungeon        NaN            Creature   
119  Venture into the dungeon, Vigilance  Legendary            Creature   
121     Flying, Venture into the dungeon        NaN            Creature   
123     Flying, Venture into the dungeon        NaN            Creature   

           subtypes  
104        Gargoyle  
115  Human, Soldier  
119  Dragon, Knight  
121           Angel  
123            Bird

In [30]:
creatures.keywords.replace(to_replace='Venture into the dungeon', value='Venture', regex=True, inplace=True)

In [31]:
creatures[creatures.keywords.str.contains('Venture', case=False)].head()

name colors  manaValue  power  toughness  \
104         Cloister Gargoyle      W        3.0    0.0        4.0   
115         Keen-Eared Sentry      W        2.0    2.0        1.0   
119  Nadaar, Selfless Paladin      W        3.0    3.0        3.0   
121               Planar Ally      W        5.0    3.0        3.0   
123             Ranger's Hawk      W        1.0    1.0        1.0   

                                                  text    rarity  \
104  When Cloister Gargoyle enters the battlefield,...  uncommon   
115  You have hexproof. (You can't be the target of...  uncommon   
119  Vigilance\nWhenever Nadaar, Selfless Paladin e...      rare   
121  Flying\nWhenever Planar Ally attacks, venture ...    common   
123  Flying\n{3}, {T}, Tap another untapped creatur...    common   

               keywords supertypes               types        subtypes  
104             Venture        NaN  Artifact, Creature        Gargoyle  
115             Venture        NaN            Creature  Human, Soldier  
119  Venture, Vigilance  Legendary            Creature  Dragon, Knight  
121     Flying, Venture        NaN            Creature           Angel  
123     Flying, Venture        NaN            Creature            Bird

In [32]:
creatures = pd.concat([creatures, creatures.keywords.str.get_dummies(sep=", ")], axis=1)

In [33]:
creatures.head()

name colors  manaValue  power  toughness  \
0       Aeronaut Admiral      W        4.0    3.0        1.0   
1       Aether Inspector      W        4.0    2.0        3.0   
2      Aethergeode Miner      W        2.0    3.0        1.0   
3      Airdrop Aeronauts      W        5.0    4.0        3.0   
4  Audacious Infiltrator      W        2.0    3.0        1.0   

                                                text    rarity  \
0          Flying\nVehicles you control have flying.  uncommon   
1  Vigilance\nWhen Aether Inspector enters the ba...    common   
2  Whenever Aethergeode Miner attacks, you get {E...      rare   
3  Flying\nRevolt — When Airdrop Aeronauts enters...  uncommon   
4  Audacious Infiltrator can't be blocked by arti...    common   

         keywords supertypes     types  ... Support  Surge  Surveil  Training  \
0          Flying        NaN  Creature  ...       0      0        0         0   
1       Vigilance        NaN  Creature  ...       0      0        0         0   
2                        NaN  Creature  ...       0      0        0         0   
3  Flying, Revolt        NaN  Creature  ...       0      0        0         0   
4                        NaN  Creature  ...       0      0        0         0   

   Trample  Transform  Undergrowth  Venture  Vigilance  Ward  
0        0          0            0        0          0     0  
1        0          0            0        0          1     0  
2        0          0            0        0          0     0  
3        0          0            0        0          0     0  
4        0          0            0        0          0     0  

[5 rows x 95 columns]

In [34]:
creatures.shape

(3811, 95)

In [35]:
creatures.columns

Index(['name', 'colors', 'manaValue', 'power', 'toughness', 'text', 'rarity',
       'keywords', 'supertypes', 'types', 'subtypes', 'Adamant', 'Adapt',
       'Afflict', 'Afterlife', 'Alliance', 'Amass', 'Ascend', 'Blitz', 'Boast',
       'Changeling', 'Channel', 'Cohort', 'Companion', 'Connive',
       'Constellation', 'Converge', 'Convoke', 'Coven', 'Cycling', 'Daybound',
       'Deathtouch', 'Defender', 'Delirium', 'Devoid', 'Disturb',
       'Double strike', 'Embalm', 'Emerge', 'Enrage', 'Equip', 'Escape',
       'Eternalize', 'Exert', 'Exploit', 'Explore', 'Fabricate', 'Fight',
       'First strike', 'Flash', 'Flying', 'Foretell', 'Haste', 'Hexproof',
       'Hexproof from', 'Improvise', 'Indestructible', 'Ingest', 'Investigate',
       'Kicker', 'Landfall', 'Learn', 'Lifelink', 'Madness', 'Magecraft',
       'Meld', 'Menace', 'Mentor', 'Mill', 'Mutate', 'Nightbound', 'Ninjutsu',
       'Pack tactics', 'Proliferate', 'Protection', 'Prowess', 'Raid', 'Rally',
       'Reach', 'Recon

In [36]:
creatures.types.unique()

array(['Creature', 'Artifact, Creature', 'Enchantment, Creature'],
      dtype=object)

In [37]:
creatures = pd.concat([creatures, creatures.types.str.get_dummies(sep=", ")], axis=1)

In [38]:
creatures.head()

name colors  manaValue  power  toughness  \
0       Aeronaut Admiral      W        4.0    3.0        1.0   
1       Aether Inspector      W        4.0    2.0        3.0   
2      Aethergeode Miner      W        2.0    3.0        1.0   
3      Airdrop Aeronauts      W        5.0    4.0        3.0   
4  Audacious Infiltrator      W        2.0    3.0        1.0   

                                                text    rarity  \
0          Flying\nVehicles you control have flying.  uncommon   
1  Vigilance\nWhen Aether Inspector enters the ba...    common   
2  Whenever Aethergeode Miner attacks, you get {E...      rare   
3  Flying\nRevolt — When Airdrop Aeronauts enters...  uncommon   
4  Audacious Infiltrator can't be blocked by arti...    common   

         keywords supertypes     types  ... Training  Trample  Transform  \
0          Flying        NaN  Creature  ...        0        0          0   
1       Vigilance        NaN  Creature  ...        0        0          0   
2                        NaN  Creature  ...        0        0          0   
3  Flying, Revolt        NaN  Creature  ...        0        0          0   
4                        NaN  Creature  ...        0        0          0   

   Undergrowth  Venture  Vigilance  Ward  Artifact  Creature  Enchantment  
0            0        0          0     0         0         1            0  
1            0        0          1     0         0         1            0  
2            0        0          0     0         0         1            0  
3            0        0          0     0         0         1            0  
4            0        0          0     0         0         1            0  

[5 rows x 98 columns]

In [39]:
creatures.shape

(3811, 98)

Note to self:  Creature column should be 100% 1.  Probably need to drop it.

In [40]:
creatures.supertypes.unique()

array([nan, 'Legendary', 'Snow', 'Legendary, Snow'], dtype=object)

In [41]:
creatures = pd.concat([creatures, creatures.supertypes.str.get_dummies(sep=", ")], axis=1)

In [42]:
creatures.shape

(3811, 100)

In [43]:
creatures.subtypes.unique()

array(['Human, Pilot', 'Dwarf, Artificer', 'Dwarf, Scout', 'Dwarf, Rogue',
       'Dwarf, Soldier', 'Bird', 'Dwarf, Warrior', 'Angel', 'Cat, Beast',
       'Dwarf, Advisor', 'Vedalken, Artificer', 'Whale', 'Human, Wizard',
       'Drake', 'Human, Artificer', 'Crab', 'Vedalken, Rogue', 'Fish',
       'Human, Pirate', 'Aetherborn, Rogue', 'Aetherborn, Artificer',
       'Insect', 'Human, Rogue', 'Aetherborn, Vampire', 'Demon',
       'Aetherborn, Warrior', 'Human, Warrior', 'Giant', 'Dragon',
       'Lizard', 'Gremlin', 'Elf, Artificer, Druid', 'Cat', 'Elf, Druid',
       'Elephant', 'Elf, Warrior', 'Human, Druid', 'Pangolin, Beast',
       'Cat, Monkey', 'Elf, Archer', 'Boar', 'Dwarf, Pilot',
       'Human, Soldier', 'Snake', 'Construct', 'Juggernaut',
       'Assembly-Worker', 'Thopter', 'Shapeshifter', 'Elf, Soldier',
       'Golem', 'Pegasus', 'Dog', 'Unicorn', 'Gargoyle', 'Human, Cleric',
       'Orc, Knight', 'Human, Monk', 'Jellyfish', 'Dwarf, Ranger',
       'Spirit, Knight', 'Hu

In [44]:
creatures = pd.concat([creatures, creatures.subtypes.str.get_dummies(sep=", ")], axis=1)

In [45]:
creatures.shape

(3811, 291)

In [46]:
creatures.head()

name colors  manaValue  power  toughness  \
0       Aeronaut Admiral      W        4.0    3.0        1.0   
1       Aether Inspector      W        4.0    2.0        3.0   
2      Aethergeode Miner      W        2.0    3.0        1.0   
3      Airdrop Aeronauts      W        5.0    4.0        3.0   
4  Audacious Infiltrator      W        2.0    3.0        1.0   

                                                text    rarity  \
0          Flying\nVehicles you control have flying.  uncommon   
1  Vigilance\nWhen Aether Inspector enters the ba...    common   
2  Whenever Aethergeode Miner attacks, you get {E...      rare   
3  Flying\nRevolt — When Airdrop Aeronauts enters...  uncommon   
4  Audacious Infiltrator can't be blocked by arti...    common   

         keywords supertypes     types  ... Weird  Werewolf  Whale  Wizard  \
0          Flying        NaN  Creature  ...     0         0      0       0   
1       Vigilance        NaN  Creature  ...     0         0      0       0   
2                        NaN  Creature  ...     0         0      0       0   
3  Flying, Revolt        NaN  Creature  ...     0         0      0       0   
4                        NaN  Creature  ...     0         0      0       0   

   Wolf  Wolverine  Worm  Wurm  Yeti  Zombie  
0     0          0     0     0     0       0  
1     0          0     0     0     0       0  
2     0          0     0     0     0       0  
3     0          0     0     0     0       0  
4     0          0     0     0     0       0  

[5 rows x 291 columns]

In [47]:
creatures = pd.concat([creatures, creatures.colors.str.get_dummies(sep=", ")], axis=1)

In [48]:
creatures.shape

(3811, 297)

In [49]:
creatures.colors.unique()

array(['W', 'U', 'B', 'R', 'G', 'R, U', 'G, R', 'G, W', 'R, W', 'G, U',
       'U, W', 'B, R', 'B, G', 'C', 'B, W', 'B, U', 'G, R, W',
       'B, G, R, U, W', 'B, G, R', 'R, U, W', 'B, G, U', 'W, G',
       'G, R, U', 'B, G, W', 'B, R, W', 'G, U, W', 'B, U, W', 'B, R, U',
       'G, R, U, W'], dtype=object)

In [50]:
creatures_transformed = creatures.drop(columns=['colors', 'text', 'rarity', 'keywords', 'supertypes', 'types', 'subtypes'])

In [51]:
creatures_transformed.head()

name  manaValue  power  toughness  Adamant  Adapt  \
0       Aeronaut Admiral        4.0    3.0        1.0        0      0   
1       Aether Inspector        4.0    2.0        3.0        0      0   
2      Aethergeode Miner        2.0    3.0        1.0        0      0   
3      Airdrop Aeronauts        5.0    4.0        3.0        0      0   
4  Audacious Infiltrator        2.0    3.0        1.0        0      0   

   Afflict  Afterlife  Alliance  Amass  ...  Worm  Wurm  Yeti  Zombie  B  C  \
0        0          0         0      0  ...     0     0     0       0  0  0   
1        0          0         0      0  ...     0     0     0       0  0  0   
2        0          0         0      0  ...     0     0     0       0  0  0   
3        0          0         0      0  ...     0     0     0       0  0  0   
4        0          0         0      0  ...     0     0     0       0  0  0   

   G  R  U  W  
0  0  0  0  1  
1  0  0  0  1  
2  0  0  0  1  
3  0  0  0  1  
4  0  0  0  1  

[5 rows x 290 columns]

In [52]:
creatures_transformed.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3811 entries, 0 to 3810
Data columns (total 290 columns):
 #    Column           Dtype  
---   ------           -----  
 0    name             object 
 1    manaValue        float64
 2    power            float64
 3    toughness        float64
 4    Adamant          int64  
 5    Adapt            int64  
 6    Afflict          int64  
 7    Afterlife        int64  
 8    Alliance         int64  
 9    Amass            int64  
 10   Ascend           int64  
 11   Blitz            int64  
 12   Boast            int64  
 13   Changeling       int64  
 14   Channel          int64  
 15   Cohort           int64  
 16   Companion        int64  
 17   Connive          int64  
 18   Constellation    int64  
 19   Converge         int64  
 20   Convoke          int64  
 21   Coven            int64  
 22   Cycling          int64  
 23   Daybound         int64  
 24   Deathtouch       int64  
 25   Defender         int64  
 26   Delirium         i

In [53]:
creatures_transformed.describe()

manaValue        power    toughness      Adamant        Adapt  \
count  3811.000000  3811.000000  3811.000000  3811.000000  3811.000000   
mean      3.425872     2.671740     2.893466     0.001837     0.003149   
std       1.595746     1.592168     1.598453     0.042824     0.056033   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       2.000000     2.000000     2.000000     0.000000     0.000000   
50%       3.000000     2.000000     3.000000     0.000000     0.000000   
75%       4.000000     3.000000     4.000000     0.000000     0.000000   
max      15.000000    13.000000    17.000000     1.000000     1.000000   

           Afflict    Afterlife     Alliance        Amass       Ascend  ...  \
count  3811.000000  3811.000000  3811.000000  3811.000000  3811.000000  ...   
mean      0.002362     0.002624     0.002624     0.002362     0.003674  ...   
std       0.048545     0.051164     0.051164     0.048545     0.060507  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
50%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
75%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
max       1.000000     1.000000     1.000000     1.000000     1.000000  ...   

              Worm         Wurm         Yeti       Zombie            B  \
count  3811.000000  3811.000000  3811.000000  3811.000000  3811.000000   
mean      0.000262     0.004723     0.000262     0.045395     0.214117   
std       0.016199     0.068572     0.016199     0.208196     0.410262   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

                 C            G            R            U            W  
count  3811.000000  3811.000000  3811.000000  3811.000000  3811.000000  
mean      0.072684     0.231698     0.219365     0.196536     0.228811  
std       0.259652     0.421972     0.413871     0.397431     0.420123  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000     0.000000  
75%       0.000000     0.000000     0.000000     0.000000     0.000000  
max       1.000000     1.000000     1.000000     1.000000     1.000000  

[8 rows x 289 columns]

This is looking like we could actually start to use it for some machine learning analysis, so I'm going to export it to a CSV and pick up in a new file.

In [54]:
creatures_transformed.to_csv('cleaned_datasets/creatures_transformed.csv')